pandas/tests/groupby/test_groupby.py

```diff
@@ -2348,6 +2348,52 @@ def test_groupby_duplicate_index():
     tm.assert_series_equal(result, expected)
 
 
+def test_group_on_empty_multiindex(transformation_func, request):
+    # GH 47787
+    # With one row, those are transforms so the schema should be the same
+    if transformation_func == "tshift":
+        mark = pytest.mark.xfail(raises=NotImplementedError)
+        request.node.add_marker(mark)
+    df = DataFrame(
+        data=[[1, Timestamp("today"), 3, 4]],
+        columns=["col_1", "col_2", "col_3", "col_4"],
+    )
+    df = df.set_index(["col_1", "col_2"])
+    if transformation_func == "fillna":
+        args = ("ffill",)
+    elif transformation_func == "tshift":
+        args = (1, "D")
+    else:
+        args = ()
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(df.index, result.index)
+    has_same_columns = (
+        type(result) == DataFrame and (df.columns == result.columns).all()
+    )
+
+    col_3 = df["col_3"]
+    result = col_3.groupby(["col_1"]).transform(transformation_func, *args)
+    tm.assert_index_equal(col_3.index, result.index)
+
+    # When empty, expect the same schema as well
+    df = DataFrame(data=[], columns=["col_1", "col_2", "col_3", "col_4"], dtype=int)
+    df = df.set_index(["col_1", "col_2"])
+    result = df.groupby(["col_1"]).transform(transformation_func, *args)
+    assert df.index.names == result.index.names
+    type_changes = {"pct_change", "rank"}
+    # pct change will return a float, so different from the original type
+    if has_same_columns and transformation_func not in type_changes:
+        tm.assert_frame_equal(df, result)
+
+    col_3 = df["col_3"]
+    series_result = col_3.groupby(["col_1"]).transform(transformation_func, *args)
+    if transformation_func not in type_changes:
+        if has_same_columns and transformation_func:
+            tm.assert_series_equal(col_3, series_result)
+        else:
+            tm.assert_series_equal(series_result, result)
```

Comment:
Similar remarks here. It's helpful to always keep `result` as "the thing being tested" and `expected` as "the thing `result` should be". Many tests (but certainly not all) are written this way.

Reply:
I can make that change